builtin/logical/ssh/storage.go

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) 2024 OpenBao a Series of LF Projects, LLC
+// SPDX-License-Identifier: MPL-2.0
+
+package ssh
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/openbao/openbao/sdk/v2/helper/errutil"
+	"github.com/openbao/openbao/sdk/v2/logical"
+)
+
+const (
+	defaultRef = "default"
+
+	MigrationInfoKey    = "config/migrationInfo"
+	issuerPrefix        = "config/issuer/"
+	storageIssuerConfig = "config/issuers"
+
+	caPublicKey                       = "ca_public_key"
+	caPrivateKey                      = "ca_private_key"
+	caPublicKeyStoragePath            = "config/ca_public_key"
+	caPublicKeyStoragePathDeprecated  = "public_key"
+	caPrivateKeyStoragePath           = "config/ca_private_key"
+	caPrivateKeyStoragePathDeprecated = "config/ca_bundle"
+
+	maxRolesToScanOnIssuerChange = 100
+	maxRolesToFindOnIssuerChange = 10
+
+	// Used as a quick sanity check for a reference id lookups...
+	uuidLength = 36
+)
+
+const (
+	IssuerRefNotFound = "not-found"
+)
+
+type issuerEntry struct {
+	ID         string `json:"id"`
+	Name       string `json:"name"`
+	PublicKey  string `json:"public_key"`
+	PrivateKey string `json:"private_key"`
+	Version    uint   `json:"version"`
+}
+
+type issuerConfigEntry struct {
+	DefaultIssuerID string `json:"default"`
+}
+
+type storageContext struct {
+	Context context.Context
+	Storage logical.Storage
+	Backend *backend
+}
+
+// makeStorageContext creates a storage context with the given context and storage
+func (b *backend) makeStorageContext(ctx context.Context, s logical.Storage) *storageContext {
+	return &storageContext{
+		Context: ctx,
+		Storage: s,
+		Backend: b,
+	}
+}
+
+// writeIssuer writes an issuerEntry to storage
+func (sc *storageContext) writeIssuer(issuer *issuerEntry) error {
```

Comment:
@Gabrielopesantos On the call today, we noticed that we need an `importIssuer(...)` so that if you call the `config/ca` endpoint multiple times with the same key material, it won't endlessly create new issuers with the same key material, but instead point out that it already exists (single entry).

Reply:
What if the key material is from an issuer that is not set as default and when one being imported is being set as default. Do we want to set the existing issuer as the default?

Reply:
If there's an existing default, I'd leave it alone in `importIssuer(...)` and only set one if there's no issues at all. Then the `config/ca` can do the write always and `issuers/import/<name>` can do it conditionally on the `set_default` value.